### ISSUE TITLE
Modal optimization

### ISSUE THREAD

**Issue by GillesOrban** [ANR-COMPASS/shesha#8]
Hi,

The modal gain optimization leads to unexpected results.
Running the test `test_sh_modopti.py`:
- the gain after optimization, `sim.rtc.d_control[0].gain`, is 1.0, while the maximum gain is set to 0.9 (`p_controller0.set_gmax(0.4)`)
- the Strehl is obviously very low and far from optimum with gain=1

In addition to this unexpected behavior, I get sometimes the following error message:
> Computing S2M matrix...
> /home/keckrtc/miniconda3/conda-bld/compass_1549269616175/work/libcarma/src.cpp/carma_magma.cpp@184 magma returned error 64: function-specific error, see documentation.
> Computing transfer functions...
When I get this error message, the loop hangs after 2 iterations, and I have to kill my ipython command.

Thank you for your follow-up,
Regards,
Gilles

**Comment by floferreira** [comment]
Hi Gilles,

The gain you pointed out is not use in modal optimization mode. The modal gains you are looking for are saved in the vector sim.rtc.d_control[0].d_gain

Your second point is more weird... I am not able to reproduce this behavior. What is your setup (OS, GPU) ?
I let this issue open while we investigate on this

**Comment by GillesOrban** [comment]
Hi Florian,

thank you for clarifying the first point. Sorry if this is obvious but how do I get, say a numpy array from this carma object? (print d_gain in the ipython shell return me the structure of the carma_obj not its values).

I am running COMPASS on ubuntu 16.04.3 LTS, 2 GPU GTX1080Ti (the test was run on one).
If I can pinpoint the problem a bit more precisely, I let you know. Currently, I cannot reproduce it systematically.

**Comment by floferreira** [comment]
To get a numpy array from a carma obj, you just have to create a numpy array from this obj :
A = np.array(your_carma_obj)
And that's all

Ok so your system is quite recent... Indeed, if you are able to isolate the issue to make it reproducible, it will be a great help. 

Thanks you

**Comment by GillesOrban** [comment]
ok. Sorry that was pretty obvious.

Ok, I think I have nailed the problem.
If the number of optimized gain is set to a larger number that the total number of actuators (sum of _ntoact for all dm; in the case of the test, set_nmodes(>62), one get the error I quoted earlier:
`/home/keckrtc/miniconda3/conda-bld/compass_1549269616175/work/libcarma/src.cpp/carma_magma.cpp@184 magma returned error 61: function-specific error, see documentation.`
which sometimes stalls, sometimes not my system.
So a beginner mistake (mine) at the root at this problem.

Thanks again for your follow-up.
